caddy/app.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package caddy provides an app and handler for Caddy Server (https://caddyserver.com/)
+// allowing it to turn any handler into one supporting the Vulcain protocol.
+
+package caddy
+
+import (
+	"errors"
+
+	outline_prometheus "github.com/Jigsaw-Code/outline-ss-server/prometheus"
+	outline "github.com/Jigsaw-Code/outline-ss-server/service"
+	"github.com/caddyserver/caddy/v2"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+)
+
+func init() {
+	caddy.RegisterModule(OutlineApp{})
+}
+
+const moduleName = "outline"
+
+type ShadowsocksConfig struct {
+	ReplayHistory int `json:"replay_history,omitempty"`
+}
+
+type OutlineApp struct {
+	Version           string             `json:"version,omitempty"`
+	ShadowsocksConfig *ShadowsocksConfig `json:"shadowsocks,omitempty"`
+
+	ReplayCache outline.ReplayCache
+	logger      *zap.Logger
+	Metrics     outline.ServiceMetrics
+	buildInfo   *prometheus.GaugeVec
+}
+
+func (OutlineApp) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  moduleName,
+		New: func() caddy.Module { return new(OutlineApp) },
+	}
+}
+
+// Provision sets up Outline.
+func (app *OutlineApp) Provision(ctx caddy.Context) error {
+	app.logger = ctx.Logger(app)
+	defer app.logger.Sync()
+
+	app.logger.Info("provisioning app instance")
+
+	if app.Version == "" {
+		app.Version = "dev"
+	}
+	if app.ShadowsocksConfig != nil {
+		app.ReplayCache = outline.NewReplayCache(app.ShadowsocksConfig.ReplayHistory)
+	}
+
+	app.defineMetrics()
```

Comment:
This should probably log error in case of error if we think it's not a fatal error.